examples/deepcopy-gen/generators/deepcopy.go

```diff
@@ -286,27 +286,76 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
 	return true
 }
 
-// hasDeepCopyMethod returns true if an appropriate DeepCopy() method is
-// defined for the given type.  This allows more efficient deep copy
+// deepCopyMethod returns the signature of a DeepCopy() method, nil or calls glog.Fatalf
+// if the type does not match. This allows more efficient deep copy
 // implementations to be defined by the type's author.  The correct signature
 // for a type T is:
 //    func (t T) DeepCopy() T
 // or:
-//    func (t *T) DeepCopy() T
-func hasDeepCopyMethod(t *types.Type) bool {
-	for mn, mt := range t.Methods {
-		if mn != "DeepCopy" {
-			continue
-		}
-		if len(mt.Signature.Parameters) != 0 {
-			return false
-		}
-		if len(mt.Signature.Results) != 1 || mt.Signature.Results[0].Name != t.Name {
-			return false
-		}
-		return true
+//    func (t *T) DeepCopy() *T
+func deepCopyMethod(t *types.Type) *types.Signature {
+	f, found := t.Methods["DeepCopy"]
+	if !found {
+		return nil
 	}
-	return false
+	if len(f.Signature.Parameters) != 0 {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected no parameters", t)
+	}
+	if len(f.Signature.Results) != 1 {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected exactly one result", t)
+	}
+
+	ptrResult := f.Signature.Results[0].Kind == types.Pointer && f.Signature.Results[0].Elem.Name == t.Name
+	nonPtrResult := f.Signature.Results[0].Name == t.Name
+
+	if !ptrResult && !nonPtrResult {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected to return %s or *%s", t, t.Name.Name, t.Name.Name)
+	}
+
+	ptrRcvr := f.Signature.Receiver != nil && f.Signature.Receiver.Kind == types.Pointer && f.Signature.Receiver.Elem.Name == t.Name
+	nonPtrRcvr := f.Signature.Receiver != nil && f.Signature.Receiver.Name == t.Name
+
+	if ptrResult != ptrRcvr || nonPtrResult != nonPtrRcvr {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected a receiver of type %s or *%s matching the result type", t, t.Name.Name, t.Name.Name)
+	}
+
+	return f.Signature
+}
+
+// deepCopyIntoMethod returns the signature of a DeepCopyInto() method, nil or calls glog.Fatalf
+// if the type is wrong. DeepCopyInto allows more efficient deep copy
+// implementations to be defined by the type's author.  The correct signature
+// for a type T is:
+//    func (t T) DeepCopyInto(t *T)
+// or:
+//    func (t *T) DeepCopyInto(t *T)
+func deepCopyIntoMethod(t *types.Type) *types.Signature {
```

Comment:
should we perhaps demand that a user defines both of these or neither of these?  Or are we ok to mix-and-match?

Reply:
We are ok. One alone works for us. The DeepCopyInto is usually an optimization with less allocations.